packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -20,64 +23,418 @@ const double _kHandleSize = 22.0;
+        child: Material(
+          elevation: 1.0,
+          child: _TextSelectionToolbarItems(
+            isAbove: widget.isAbove,
+            overflowOpen: _overflowOpen,
+            children: <Widget>[
+              // The navButton that shows and hides the overflow menu is the
+              // first child.
+              Material(
+                child: IconButton(
+                  // TODO(justinmc): This should be an AnimatedIcon, but
+                  // AnimatedIcons doesn't yet support arrow_back to more_vert.
+                  // https://github.com/flutter/flutter/issues/51209
+                  icon: Icon(_overflowOpen ? Icons.arrow_back : Icons.more_vert),
+                  onPressed: () {
+                    setState(() {
+                      _overflowOpen = !_overflowOpen;
+                    });
+                  },
+                  tooltip: _overflowOpen ? 'Back' : 'More',
+                ),
+              ),
+              ...items,
+            ],
+          ),
+        ),
       ),
     );
+
+    return _overflowOpen ? IntrinsicWidth(child: child) : child;
+  }
+}
+
+// Renders the menu items in the correct positions in the menu and its overflow
+// submenu based on calculating which item would first overflow.
+class _TextSelectionToolbarItems extends MultiChildRenderObjectWidget {
+  _TextSelectionToolbarItems({
+    Key key,
+    @required List<Widget> children,
+    @required this.isAbove,
+    @required this.overflowOpen,
+  }) : assert(children != null),
+       assert(isAbove != null),
+       assert(overflowOpen != null),
+       super(key: key, children: children);
+
+  final bool isAbove;
+  final bool overflowOpen;
+
+  @override
+  _TextSelectionToolbarItemsRenderBox createRenderObject(BuildContext context) {
+    return _TextSelectionToolbarItemsRenderBox(
+      isAbove: isAbove,
+      overflowOpen: overflowOpen,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _TextSelectionToolbarItemsRenderBox renderObject) {
+    renderObject
+      ..isAbove = isAbove
+      ..overflowOpen = overflowOpen;
+  }
+
+  @override
+  MultiChildRenderObjectElement createElement() => MultiChildRenderObjectElement(this);
+}
+
+class _TextSelectionToolbarItemsRenderBox extends RenderBox with ContainerRenderObjectMixin<RenderBox, FlexParentData>, RenderBoxContainerDefaultsMixin<RenderBox, FlexParentData> {
+  _TextSelectionToolbarItemsRenderBox({
+    @required bool isAbove,
+    @required bool overflowOpen,
+  }) : assert(overflowOpen != null),
+       assert(isAbove != null),
+       _isAbove = isAbove,
+       _overflowOpen = overflowOpen,
+       super();
+
+  // This is static so that it can be preserved between instances. The
+  // IntrinsicWidth widget is added and removed above this in the tree, causing
+  // it to be recreated, but _intrinsicWidth is needed immediately when the
+  // overflow menu is open.
+  static double _intrinsicWidth;
```

Comment:
Caching this seems dangerous. Can we not just calculate this on-demand when getIntrinsicsWidth is called (by calling getInstrinsicsWidth on them)?

Reply:
The problem is that at the time I'm asked for intrinsic width, I don't yet have `constraints`, so I can't know where the menu will overflow.  I can calculate the true intrinsic width of the menu without any constraints, but that doesn't help our IntrinsicWidth widget hack.

Is there any other way I can do it? A hack that comes to mind is to use the width of the screen instead of constraints.maxWidth...